### PR TITLE
Stop AIs from shooting at dead players + more

### LINKED ISSUE #533: Make AIs only prioritize each other when no real players

### PR DESCRIPTION
**Pull request by toberge**
Also closes #533

### CODE REVIEW

File: Assets/Scripts/Gamestate/AIManager.cs [hunk 1]
```diff
@@ -195,32 +210,20 @@ private IEnumerator LookForTargets()
...
+        var closestDistance = ignoreAwareRadius;
 
-            closestPlayer = player.AiTarget.transform;
-            closestDistance = hitDistance;
-            DestinationTarget = closestPlayer;
-            ShootingTarget = player.AiAimSpot;
+        // TODO the ammo box stuff!
+        var hasAmmoBoxBody = ammoBoxCollector.CanReload;
+        var isOutOfAmmo = gunController && gunController.stats.Ammo < 1;
+        if (hasAmmoBoxBody && isOutOfAmmo)
```
Comment (Fueredoriku) [inline Assets/Scripts/Gamestate/AIManager.cs]: This should also be called when the AI is in close proximity mode! If the AI depletes its ammo and is in close range it'll soft lock itself not being able to shoot and unwilling to find new ammo

File: Assets/Scripts/Gamestate/AIManager.cs [hunk 2]
```diff
@@ -237,10 +240,23 @@ private void FindPlayers()
             aiMovement.Target = ShootingTarget;
             var isStrafeDistance = closestDistance < 10f;
             aiMovement.enabled = isStrafeDistance;
+
+            var closestExplodingBarrel = ExplodingBarrel.GetViableExplodingBarrel(ShootingTarget.position);
+            var isBarrelSafeToShoot = closestExplodingBarrel
+                && Vector3.Distance(closestExplodingBarrel.transform.position, transform.position) > closestExplodingBarrel.Radius + 1;
+            if (isBarrelSafeToShoot)
+            {
+                Debug.Log($"Found exploding barrel at {closestExplodingBarrel.transform.position} >:)", closestExplodingBarrel);
```
Comment (Fueredoriku) [inline Assets/Scripts/Gamestate/AIManager.cs]: remove debug log?  🙏

**Review by Fueredoriku — CHANGES_REQUESTED** [review]
Great work, the AIs become suprisingly more fun to play against when they can use seeker and tether too!
 
I have identified three issues that need fixing:
- Ammoboxes ignored if AI has entered a close range state
- Tether may never be replugged
(I suggest we fix the tether problem by forcing the AIs to always aim straight down when planting the plug)
tethersoftlock



- Also it seems that if the player is gone, the AIs may sometimes choose to stand still at some place and shoot the remaining targets from that single spot only. To replicate this problem, make one or more of the AIs target you, die, then spectate and there is a good chance that they may get stuck standing in a specific place. See one of the comments I made for an example.

The rest looks good here, so if we fix these issues this should be good to go!

File: Assets/Scripts/Gamestate/AIManager.cs [hunk 3]
```diff
@@ -260,6 +276,55 @@ private void FindPlayers()
...
+            if (areOnlyAiPlayersLeft)
+            {
+                if (hitDistance > closestDistance)
+                    continue;
+
+                closestDistance = hitDistance;
+                closestPlayer = player.AiTarget.transform;
+                ShootingTarget = player.AiAimSpot;
+                continue;
+            }
```
Comment (Fueredoriku) [inline Assets/Scripts/Gamestate/AIManager.cs:339]: There seems to be a problem with something inside the areOnlyAiPlayersLeft, making it so the AIs can choose to stand still firing at eachother from places they can't hit eachother, thus softlocking the game

Here is an example spot from The Pit
image

File: Assets/Scripts/Gamestate/AIManager.cs (hunk 2 above)
Comment (toberge) [reply Assets/Scripts/Gamestate/AIManager.cs]: Removed 😔

File: Assets/Scripts/Gamestate/AIManager.cs (hunk 1 above)
Comment (toberge) [reply Assets/Scripts/Gamestate/AIManager.cs]: Reworked so it happens when they are in close proximity, _without_ messing up their trajectory if they are in the air

File: Assets/Scripts/Gamestate/AIManager.cs (hunk 3 above)
Comment (toberge) [reply Assets/Scripts/Gamestate/AIManager.cs:339]: Fixed. Removed the ShootingTarget and continue; lines, instead setting DestinationTarget.

**Comment by toberge** [comment]
> Great work, the AIs become suprisingly more fun to play against when they can use seeker and tether too!
> 
> I have identified three issues that need fixing:
> 
> * Ammoboxes ignored if AI has entered a close range state

Fixed, see other comment

>  * Tether may never be replugged
>     (I suggest we fix the tether problem by forcing the AIs to always aim straight down when planting the plug)

Fixed as suggested


> * Also it seems that if the player is gone, the AIs may sometimes choose to stand still at some place and shoot the remaining targets from that single spot only. To replicate this problem, make one or more of the AIs target you, die, then spectate and there is a good chance that they may get stuck standing in a specific place. See one of the comments I made for an example.

Fixed. They don't instantly set ShootingTarget here anymore, instead DestinationTarget so that they walk to their enemy first.

File: Assets/Scripts/Gamestate/AIManager.cs [hunk 4]
```diff
@@ -186,61 +227,78 @@ private IEnumerator LookForTargets()
...
+                var closestExplodingBarrel = ExplodingBarrel.GetViableExplodingBarrel(ShootingTarget.position);
+                var isBarrelSafeToShoot = closestExplodingBarrel
+                    && Vector3.Distance(closestExplodingBarrel.transform.position, transform.position) > closestExplodingBarrel.Radius + 1;
+                if (isBarrelSafeToShoot)
+                {
+                    // TODO respond to the barrel actually exploding?
+                    // TODO the shooting target going missing seems to be the cause for 
+                    ShootingTarget = closestExplodingBarrel.transform;
+                    // TODO unsub?
+                    closestExplodingBarrel.GetComponent<HealthController>().onDeath += HandleExplodedBarrel;
```
Comment (Fueredoriku) [inline Assets/Scripts/Gamestate/AIManager.cs:289]: Maybe unsubscribe at the HandleExplodedBarrel function itself? Might be nice in order to minimize memory leakage

**Review by Fueredoriku — APPROVED** [review]
Great work!! 👍

The AIs seem a lot more... sane now 😅

I left a small comment but other than that I have no more feedback, so I'll leave an approve for you to use at your discretion

File: Assets/Scripts/Gamestate/AIManager.cs (hunk 1 above)
Comment (Fueredoriku) [reply Assets/Scripts/Gamestate/AIManager.cs]: Nice!

File: Assets/Scripts/Gamestate/AIManager.cs (hunk 4 above)
Comment (toberge) [reply Assets/Scripts/Gamestate/AIManager.cs:289]: Thought I did that already but must have forgor 💀

Comment (toberge) [reply Assets/Scripts/Gamestate/AIManager.cs:289]: Nope, I already did every single thing that the TODO comments around here were talking about. Removed the superfluous comments now 🙃

**Pull request merged**